documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1189 @@
+
+# Geometric Types
+
+The following Geometric Types are defined:
+- Box2F
+- Box2I
+- Box3F
+- Box3I
+- CircleF
+- CircleI
+- CubeF
+- CubeI
+- PlaneF
+- PlaneI
+- Ray2F
+- Ray2I
+- Ray3F
+- Ray3I
+- RectangleF
+- RectangleI
+- SphereF
+- SphereI
+
+The Box structs are defined by Min and Max Vectors, while Rectangle and Cube are defined by Origin and Size Vectors. These Types **must** be implicitly castable between each other.
+
+Each type **must** include the following:
+- Intersect functions with both another instance of the type and a point
+- GetDistanceToNearest(Point,Edge,etc) functions **must** be available for a given point
+- For all but the rays and planes, GetInflated function that takes a point and returns the scaled object that is closest to the original and contains the given point
+- Include Scale and Translation transformation functions
+- For Planes and Rays, Normalize functions
+- For Planes include the following static functions
+  - CreateFromVerticies
+  - Dot
+    - with a Vector4
+  - DotCoordinate and DotNormal
+    - with a Vector3
+  - Transform
+    - With a Matrix4x4 or Quaternion, if relevant
+
+# Vector Interface Definitions
+
+The following interfaces are defined for Vectors:
+
+```cs
+public interface IVector
+{
+    /// <summary>
+    /// The type of scalars in this vector type.
+    /// </summary>
+    protected static abstract Type ScalarType { get; }
+    
+    /// <summary>
+    /// The amount of scalars in this vector type.
+    /// </summary>
+    protected internal static abstract int Count { get; }
+}
+
+public interface IVectorEquatable<TVector, TScalar>
+    : IEquatable<TVector>
+    where TVector : IVector<TVector, TScalar>
+    where TScalar : INumberBase<TScalar>
+{
+    public TVector ScalarsEqual(TVector other);
+}
+
+public interface IVector<TVector, TScalar> :
+    IVector,
+    IAdditionOperators<TVector, TVector, TVector>,
+    IAdditiveIdentity<TVector, TVector>,
+    IDecrementOperators<TVector>,
+    IDivisionOperators<TVector, TVector, TVector>,
+    IEqualityOperators<TVector, TVector, bool>,
+    IIncrementOperators<TVector>,
+    IMultiplicativeIdentity<TVector, TVector>,
+    IMultiplyOperators<TVector, TVector, TVector>,
+    ISpanFormattable,
+    ISpanParsable<TVector>,
+    ISubtractionOperators<TVector, TVector, TVector>,
+    IUnaryPlusOperators<TVector, TVector>,
+    IUnaryNegationOperators<TVector, TVector>,
+    IUtf8SpanFormattable,
+    IUtf8SpanParsable<TVector>,
+    IReadOnlyList<TScalar>,
+    IVectorEquatable<TVector, TScalar>
+    where TVector : IVector<TVector, TScalar>
+    where TScalar : INumberBase<TScalar>
+{
+    static Type IVector.ScalarType => typeof(TScalar);
+
+    static virtual TVector Zero => TVector.Create(TScalar.Zero);
+
+    static virtual TVector One => TVector.Create(TScalar.One);
+
+    int IReadOnlyCollection<TScalar>.Count => TVector.Count;
+
+    TScalar IReadOnlyList<TScalar>.this[int index];
+    
+    public new TScalar this[int index];
```

Comment:
why do you need this, shouldn't it coming from `IReadOnlyList` be an identical signature?